File: scraper/dse_live.py
#!/usr/bin/env python3
"""
DSE Live Data Scraper for Personal Investment
Free solution using DSE website data
"""

import requests
from bs4 import BeautifulSoup
import pandas as pd
from datetime import datetime, date
import json
import time
from pathlib import Path

class DSELiveData:
    """Scrape real DSE data for personal investing"""

    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        })
        # Disable SSL verification for DSE (they have certificate issues)
        self.session.verify = False
        # Suppress SSL warnings
        import urllib3
        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)

    def get_current_prices(self):
        """Get latest prices from DSE website"""
        url = "https://www.dsebd.org/latest_share_price_scroll_l.php"

        try:
            response = self.session.get(url, timeout=10)
            soup = BeautifulSoup(response.content, 'html.parser')

            # Parse the price table
            table = soup.find('table', class_='table')
            if not table:
                # Try alternative URL
                return self.get_prices_alternative()

            stocks_data = {}
            rows = table.find_all('tr')[1:]  # Skip header

            for row in rows:
                cols = row.find_all('td')
                if len(cols) >= 8:
                    ticker = cols[1].text.strip()
                    stocks_data[ticker] = {
                        'last_price': float(cols[2].text.replace(',', '')),
                        'high': float(cols[3].text.replace(',', '')),
                        'low': float(cols[4].text.replace(',', '')),
                        'close': float(cols[5].text.replace(',', '')),
                        'change': float(cols[6].text.replace(',', '')),
                        'volume': int(cols[8].text.replace(',', ''))
                    }

            return stocks_data

        except Exception as e:
            print(f"DSE website error: {e}")
            print("Using simulated prices for demonstration...")
            return self.get_simulated_prices()

    def get_prices_alternative(self):
        """Try alternative data source"""
        try:
            # Alternative DSE endpoint
            url = "https://www.dsebd.org/dse_close_price.php"
            response = self.session.get(url, timeout=10)
            # Parse alternative format
            return {}
        except:
            return self.get_simulated_prices()

    def get_simulated_prices(self):
        """Generate realistic prices based on your portfolio for testing"""
        import random

        # Your actual portfolio stocks with realistic prices
        base_prices = {
            'GP': 239.7,
            'SQURPHARMA': 212.8,
            'BRACBANK': 64.1,
            'ROBI': 28.8,
            'BATBC': 370.7,
            'CITYBANK': 28.5,
            'RENATA': 900.0,
            'OLYMPIC': 140.0,
            'BERGERPBL': 1650.0,
            'MARICO': 2200.0,
            'BSCPLC': 156.3,
            'BXPHARMA': 124.0,
            'BSC': 102.4,
            'BSRMSTEEL': 78.1,
            'EBL': 25.5,
            'LHB': 52.7,
            'MIDLANDBNK': 17.3,
            'PRIMEBANK': 29.5,
            'PUBALIBANK': 33.9,
            'UPGDCL': 118.0,
            'WALTONHIL': 386.9
        }

        stocks_data = {}
        for ticker, base_price in base_prices.items():
            # Simulate daily movement (±3%)
            change_pct = random.uniform(-0.03, 0.03)
            last_price = base_price * (1 + change_pct)

            stocks_data[ticker] = {
                'last_price': round(last_price, 2),
                'high': round(last_price * 1.01, 2),
                'low': round(last_price * 0.99, 2),
                'close': round(base_price, 2),
                'change': round(last_price - base_price, 2),
                'volume': random.randint(10000, 1000000)
            }

        return stocks_data

    def get_historical_data(self, ticker: str, days: int = 30):
        """Get historical data from DSE (limited availability)"""
        # DSE provides limited historical data
        # You might need to accumulate daily data over time

        url = f"https://www.dsebd.org/php_graph/monthly_graph.php"
        params = {
            'inst': ticker,
            'duration': days,
            'type': 'price'
        }

        try:
            response = self.session.get(url, params=params, timeout=10)
            # Parse CSV or JSON response
            return self._parse_historical(response.text)
        except:
            return []

    def get_company_info(self, ticker: str):
        """Get company fundamentals"""
        url = f"https://www.dsebd.org/companylistbyindustry.php?industry={ticker}"

        try:
            response = self.session.get(url, timeout=10)
            soup = BeautifulSoup(response.content, 'html.parser')

            # Extract PE, EPS, etc from company page
            info = {
                'pe': self._extract_number(soup, 'P/E'),
                'eps': self._extract_number(soup, 'EPS'),
                'nav': self._extract_number(soup, 'NAV'),
                'market_cap': self._extract_number(soup, 'Market Cap')
            }
            return info
        except:
            return {}

    def _extract_number(self, soup, label):
        """Helper to extract numeric values"""
        try:
            element = soup.find(text=label)
            if element:
                value = element.parent.find_next_sibling().text
                return float(value.replace(',', ''))
        except:
            pass
        return 0.0

    def _parse_historical(self, data):
        """Parse historical data response"""
        lines = data.strip().split('\n')
        result = []

        for line in lines[1:]:  # Skip header
            parts = line.split(',')
            if len(parts) >= 6:
                result.append({
                    'date': parts[0],
                    'open': float(parts[1]),
                    'high': float(parts[2]),
                    'low': float(parts[3]),
                    'close': float(parts[4]),
                    'volume': int(parts[5])
                })

        return result

    def save_to_csv(self, ticker: str, data: dict):
        """Save data in MCP-compatible format"""
        output_dir = Path("data/csv")
        output_dir.mkdir(parents=True, exist_ok=True)

        # Save OHLCV
        if 'ohlcv' in data:
            df = pd.DataFrame(data['ohlcv'])
            df.to_csv(output_dir / f"{ticker}_ohlcv.csv", index=False)

        # Save fundamentals
        if 'fundamentals' in data:
            df = pd.DataFrame([data['fundamentals']])
            df.to_csv(output_dir / f"{ticker}_fundamentals.csv", index=False)

    def update_watchlist(self, tickers: list):
        """Update data for your watchlist"""
        current_prices = self.get_current_prices()

        for ticker in tickers:
            print(f"Updating {ticker}...")

            # Get current price
            if ticker in current_prices:
                price_data = current_prices[ticker]

                # Get historical if available
                historical = self.get_historical_data(ticker)

                # Get fundamentals
                fundamentals = self.get_company_info(ticker)

                # Combine and save
                data = {
                    'ohlcv': historical or [price_data],
                    'fundamentals': fundamentals
                }

                self.save_to_csv(ticker, data)
                print(f"✅ {ticker} updated")
            else:
                print(f"⚠️ {ticker} not found in current prices")

            time.sleep(1)  # Be polite to DSE servers

if __name__ == "__main__":
    # Your personal watchlist
    MY_WATCHLIST = [
        "GP",
        "SQURPHARMA",
        "BRACBANK",
        "ROBI",
        "BATBC",
        "CITYBANK",
        "RENATA",
        "OLYMPIC",
        "BERGERPBL",
        "MARICO"
    ]

    scraper = DSELiveData()
    scraper.update_watchlist(MY_WATCHLIST)
    print("\n✅ Watchlist updated successfully!")

File: packages/scraper/src/dse-live.ts
import * as cheerio from 'cheerio';
import { join } from 'node:path';
import {
  ensureDir,
  fetchText,
  parseHistoricalCsv,
  sleep,
  writeCsv,
  type OhlcvRow,
} from './utils.js';

export interface LivePriceData {
  last_price: number;
  high: number;
  low: number;
  close: number;
  change: number;
  volume: number;
}

const BASE_PRICES: Record<string, number> = {
  GP: 239.7,
  SQURPHARMA: 212.8,
  BRACBANK: 64.1,
  ROBI: 28.8,
  BATBC: 370.7,
  CITYBANK: 28.5,
  RENATA: 900.0,
  OLYMPIC: 140.0,
  BERGERPBL: 1650.0,
  MARICO: 2200.0,
  BSCPLC: 156.3,
  BXPHARMA: 124.0,
  BSC: 102.4,
  BSRMSTEEL: 78.1,
  EBL: 25.5,
  LHB: 52.7,
  MIDLANDBNK: 17.3,
  PRIMEBANK: 29.5,
  PUBALIBANK: 33.9,
  UPGDCL: 118.0,
  WALTONHIL: 386.9,
};

export class DSELiveData {
  private readonly headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
  };

  async getCurrentPrices(): Promise<Record<string, LivePriceData>> {
    const url = 'https://www.dsebd.org/latest_share_price_scroll_l.php';

    try {
      const response = await fetch(url, { headers: this.headers });
      const html = await response.text();
      const $ = cheerio.load(html);

      const table = $('table.table');
      if (!table.length) {
        return this.getPricesAlternative();
      }

      const stocksData: Record<string, LivePriceData> = {};
      const rows = table.find('tr').slice(1);

      rows.each((_, row) => {
        const cols = $(row).find('td');
        if (cols.length >= 8) {
          const ticker = $(cols[1]).text().trim();
          stocksData[ticker] = {
            last_price: parseFloat($(cols[2]).text().replace(/,/g, '')),
            high: parseFloat($(cols[3]).text().replace(/,/g, '')),
            low: parseFloat($(cols[4]).text().replace(/,/g, '')),
            close: parseFloat($(cols[5]).text().replace(/,/g, '')),
            change: parseFloat($(cols[6]).text().replace(/,/g, '')),
            volume: parseInt($(cols[8]).text().replace(/,/g, ''), 10),
          };
        }
      });

      return stocksData;
    } catch (e) {
      console.error('DSE website error:', e);
      console.log('Using simulated prices for demonstration...');
      return this.getSimulatedPrices();
    }
  }

  async getPricesAlternative(): Promise<Record<string, LivePriceData>> {
    try {
      const url = 'https://www.dsebd.org/dse_close_price.php';
      await fetchText(url);
      return {};
    } catch {
      return this.getSimulatedPrices();
    }
  }

  getSimulatedPrices(): Record<string, LivePriceData> {
    const stocksData: Record<string, LivePriceData> = {};

    for (const [ticker, basePrice] of Object.entries(BASE_PRICES)) {
      const changePct = Math.random() * 0.06 - 0.03;
      const lastPrice = basePrice * (1 + changePct);

      stocksData[ticker] = {
        last_price: Math.round(lastPrice * 100) / 100,
        high: Math.round(lastPrice * 1.01 * 100) / 100,
        low: Math.round(lastPrice * 0.99 * 100) / 100,
        close: Math.round(basePrice * 100) / 100,
        change: Math.round((lastPrice - basePrice) * 100) / 100,
        volume: Math.floor(Math.random() * 990000) + 10000,
      };
    }

    return stocksData;
  }

  async getHistoricalData(ticker: string, days = 30): Promise<OhlcvRow[]> {
    const params = new URLSearchParams({
      inst: ticker,
      duration: String(days),
      type: 'price',
    });

    const url = `https://www.dsebd.org/php_graph/monthly_graph.php?${params.toString()}`;

    try {
      const text = await fetchText(url);
      if (text) {
        return this.parseHistorical(text);
      }
    } catch {
      // fall through
    }
    return [];
  }

  async getCompanyInfo(ticker: string): Promise<Record<string, number>> {
    const url = `https://www.dsebd.org/companylistbyindustry.php?industry=${ticker}`;

    try {
      const response = await fetch(url, { headers: this.headers });
      const html = await response.text();
      const $ = cheerio.load(html);

      return {
        pe: this.extractNumber($, 'P/E'),
        eps: this.extractNumber($, 'EPS'),
        nav: this.extractNumber($, 'NAV'),
        market_cap: this.extractNumber($, 'Market Cap'),
      };
    } catch {
      return {};
    }
  }

  extractNumber($: cheerio.CheerioAPI, label: string): number {
    try {
      const element = $('*').filter((_, el) => $(el).text().trim() === label).first();
      if (element.length) {
        const value = element.parent().next().text();
        return parseFloat(value.replace(/,/g, ''));
      }
    } catch {
      // fall through
    }
    return 0.0;
  }

  parseHistorical(data: string): OhlcvRow[] {
    return parseHistoricalCsv(data);
  }

  saveToCsv(ticker: string, data: Record<string, unknown>): void {
    const outputDir = join('data', 'csv');
    ensureDir(outputDir);

    if (data.ohlcv) {
      const rows = data.ohlcv as OhlcvRow[];
      writeCsv(
        join(outputDir, `${ticker}_ohlcv.csv`),
        rows,
        ['date', 'open', 'high', 'low', 'close', 'volume'],
      );
    }

    if (data.fundamentals) {
      const fundamentals = data.fundamentals as Record<string, unknown>;
      const keys = Object.keys(fundamentals);
      writeCsv(join(outputDir, `${ticker}_fundamentals.csv`), [fundamentals], keys);
    }
  }

  async updateWatchlist(tickers: string[]): Promise<void> {
    const currentPrices = await this.getCurrentPrices();

    for (const ticker of tickers) {
      console.log(`Updating ${ticker}...`);

      if (ticker in currentPrices) {
        const priceData = currentPrices[ticker]!;
        const historical = await this.getHistoricalData(ticker);
        const fundamentals = await this.getCompanyInfo(ticker);

        const data = {
          ohlcv: historical.length > 0 ? historical : [priceData],
          fundamentals,
        };

        this.saveToCsv(ticker, data);
        console.log(`✅ ${ticker} updated`);
      } else {
        console.log(`⚠️ ${ticker} not found in current prices`);
      }

      await sleep(1000);
    }
  }
}

export const DEFAULT_WATCHLIST = [
  'GP',
  'SQURPHARMA',
  'BRACBANK',
  'ROBI',
  'BATBC',
  'CITYBANK',
  'RENATA',
  'OLYMPIC',
  'BERGERPBL',
  'MARICO',
] as const;

export async function runDseLiveMain(): Promise<void> {
  const scraper = new DSELiveData();
  await scraper.updateWatchlist([...DEFAULT_WATCHLIST]);
  console.log('\n✅ Watchlist updated successfully!');
}
